
### <b>Análisis y predicción de datos: Football Salaries FIFA 22</b>
#### <b><i>Introducción</i></b>
El mundo del fútbol, antiguamente, no se regía mucho por el dinero: era más pasión, juego, calidad y talento. Brasil dominaba en los años 90 con el jogo bonito. Equipazos como la alemania de los 70, la naranaja mecánica del fútbol total de Cruyff. La argentina del 86 de Maradona, entre otros clubes y selecciones históricas. Los fichajes y sueldos realmente no eran del todo llamativos, aunque relevancia sí tenían. Antes de la <b>Ley Bosman</b>, los jugadores extranjeros solo eran permitidos en los clubes europeos con una <b>cantidad máxima de 3</b>. Empero, a la postre de la imposición de dicha ley, todo cambió. Fichajes exorbitantes, sueldos inimaginables y dinero por todos lados. Por mencionar un ejemplo: el fichaje de Luis Figo por el Real Madrid. O más recientemente, Paul Pogba por el United, Neymar por el PSG, entre otros. Esto ha llevado a los clubes a pagar cifras exorbitantes no solo por el fichaje, sino por el jugador en sí, para mantenerlo. Esto se debe a múltiples factores, pero estos factores y todo el análisis nace de la siguiente pregunta: <b>¿Por qué sueldos tan altos?</b>
    
En nuestro afán de responder dicha pregunta, se realiza el siguiente estudio, donde se buscará predecir cuáles son esos factores que influyen en la asignación de una masa salarial alta o baja en los jugadores: ¿De qué depende? ¿Es la expectativa? ¿Es la trayectoria? ¿Es la edad? 
Todo ello lo abordaremos durante el desarrollo del estudio. 

### <b>¿Qué usaremos?</b>
Para el desarollo del estudio, utilizaremos el archivo <b><i>dataset.csv</i></b>, en el cual se encontrará un registro de todos los datos de +19 mil jugadores, y atributos que se consideraron relevantes para anotar. A continuación se muestra el diccionario de variables:

#### <b>Diccionario de datos</b>
| Variable | Tipo | Descripción | Comentario |
|----------|------|-------------|------------|
| `age` | Numérica | Edad del jugador profesional en años; útil para analizar la madurez deportiva y la etapa de la carrera. | Sports analytics |
| `height_cm` | Numérica | Estatura del jugador en centímetros. | Sports analytics |
| `weight_kg` | Numérica | Peso del jugador en kilogramos. | Sports analytics |
| `nationality_name` | Categórica | Nacionalidad del jugador. | Sports analytics |
| `overall` | Numérica | Valoración general (Overall) del jugador. | Sports analytics |
| `potential` | Numérica | Potencial máximo estimado del jugador. | Sports analytics |
| `attacking_crossing` | Numérica | Habilidad para realizar centros. | Sports analytics |
| `attacking_finishing` | Numérica | Capacidad de definición frente al arco. | Sports analytics |
| `attacking_heading_accuracy` | Numérica | Precisión en los remates de cabeza. | Sports analytics |
| `attacking_short_passing` | Numérica | Precisión en los pases cortos. | Sports analytics |
| `attacking_volleys` | Numérica | Habilidad para ejecutar voleas. | Sports analytics |
| `skill_dribbling` | Numérica | Nivel de regate del jugador. | Sports analytics |
| `skill_curve` | Numérica | Capacidad para dar efecto al balón. | Sports analytics |
| `skill_fk_accuracy` | Numérica | Precisión en tiros libres. | Sports analytics |
| `skill_long_passing` | Numérica | Precisión en pases largos. | Sports analytics |
| `skill_ball_control` | Numérica | Control del balón. | Sports analytics |
| `movement_acceleration` | Numérica | Capacidad de aceleración. | Sports analytics |
| `movement_sprint_speed` | Numérica | Velocidad máxima de sprint. | Sports analytics |
| `movement_agility` | Numérica | Agilidad del jugador. | Sports analytics |
| `movement_reactions` | Numérica | Rapidez de reacción ante las jugadas. | Sports analytics |
| `movement_balance` | Numérica | Equilibrio corporal. | Sports analytics |
| `defending_standing_tackle` | Numérica | Eficacia en entradas de pie. | Sports analytics |
| `defending_sliding_tackle` | Numérica | Eficacia en entradas deslizantes. | Sports analytics |
| `goalkeeping_diving` | Numérica | Capacidad del portero para lanzarse. | Sports analytics |
| `goalkeeping_handling` | Numérica | Seguridad al atrapar el balón. | Sports analytics |
| `goalkeeping_kicking` | Numérica | Precisión y potencia en los despejes del portero. | Sports analytics |
| `goalkeeping_positioning` | Numérica | Posicionamiento del portero. | Sports analytics |
| `goalkeeping_reflexes` | Numérica | Reflejos del portero. | Sports analytics |
| `wage_eur` | Numérica | **Variable objetivo (Target): salario del jugador expresado en euros.** | Objetivo del modelo |

### <b><i>Fase I:</i> Análisis Exploratorio de Datos (EDA)</b>
En esta etapa tendremos 4 objetivos claros y específicos:
1. Determinar la calidad de los datos: identificar missing values, valores incoherentes, repetición de data (duplicados).
2. Descubriremos cuál es la distribución de la variable objetivo.
3. Analizaremos las variables que se consideren más relevantes para el estudio.
4. Expondremos nuestros hallazgos de manera que sea entendible en el contexto futbolístico.

#### <b><i>1. Determinación de la calidad de los datos</i></b>
Esta parte es fundamental en un EDA, ya que nos permite saber si los datos son óptimos para el correcto análisis y procesamiento. Dentro del desarrollo de esta parte, corresponde identificar lo siguiente:
- Valores missing
- Valores duplicados
- Valores incoherentes

In [1]:
"""
======== DETERMINACIÓN DE LA CALIDAD DE DATOS DEL MODELO =========
"""

# Importamos las librerías necesarias
import os
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sbs

In [2]:
# Obtenemos la ruta e importamos el dataframe
path = os.path.join(os.getcwd(), "..", "data", "raw", "dataset.csv")
df = pd.read_csv(path)
df.head(10)

,age,height_cm,weight_kg,nationality_name,overall,potential,attacking_crossing,attacking_finishing,attacking_heading_accuracy,attacking_short_passing,...,movement_reactions,movement_balance,defending_standing_tackle,defending_sliding_tackle,goalkeeping_diving,goalkeeping_handling,goalkeeping_kicking,goalkeeping_positioning,goalkeeping_reflexes,wage_eur
0,34,170,72,Argentina,93,93,85,95,70,91,...,94,95,35,24,6,11,15,14,8,320000.0
1,32,185,81,Poland,92,92,71,95,90,85,...,93,82,42,19,15,6,12,8,10,270000.0
2,36,187,83,Portugal,91,91,87,95,90,80,...,94,74,32,24,7,11,15,14,11,270000.0
3,29,175,68,Brazil,91,91,85,83,63,86,...,89,84,32,29,9,9,15,15,11,270000.0
4,30,181,70,Belgium,91,91,94,82,55,94,...,91,78,65,53,15,13,5,10,13,350000.0
5,28,188,87,Slovenia,91,93,13,11,15,43,...,88,49,12,18,87,92,78,90,90,130000.0
6,22,182,73,France,91,95,78,93,72,85,...,93,83,34,32,13,5,7,11,6,230000.0
7,35,193,93,Germany,90,90,15,13,25,60,...,87,35,10,11,88,88,91,89,88,86000.0
8,29,187,85,Germany,90,92,18,14,11,61,...,86,43,13,10,88,85,88,88,90,250000.0
9,27,188,89,England,90,90,80,94,86,85,...,92,70,36,38,8,10,11,14,11,240000.0


### <b>1. Calidad de los datos y limpieza</b>
##### <b>A. Valores missing</b>

In [3]:
# A. Limpieza de datos
# Valores missing
missing = pd.DataFrame({
    "Missing": df.isnull().sum(),
    "Porcentaje": (df.isnull().sum()/len(df))*100
})

missing

,Missing,Porcentaje
age,0,0.0
height_cm,0,0.0
weight_kg,0,0.0
nationality_name,0,0.0
overall,0,0.0
potential,0,0.0
attacking_crossing,0,0.0
attacking_finishing,0,0.0
attacking_heading_accuracy,0,0.0
attacking_short_passing,0,0.0


#### <b><i>Conclusión</i></b>
Según el dataset proporcionado y la manipulación de código llevada a cabo, se llega a la conclusión de que el dataset no posee ningún dato faltante y que cada celda está completa. Esto es un buen inidicio de que el csv. fue entregado con completitud de datos y no hará falta ajustes adicionales.

#### <b><i>Decisión</i></b>
No se modificará el dataset.

##### <b>B. Valores duplicados</b>

In [4]:
# Valores duplicados
# Cantidad de duplicados
duplicados = df.duplicated().sum()
print(f"La cantidad de duplicados es: {duplicados}")

# Comprobamos viendo los duplicados
df[df.duplicated()]

La cantidad de duplicados es: 0


,age,height_cm,weight_kg,nationality_name,overall,potential,attacking_crossing,attacking_finishing,attacking_heading_accuracy,attacking_short_passing,...,movement_reactions,movement_balance,defending_standing_tackle,defending_sliding_tackle,goalkeeping_diving,goalkeeping_handling,goalkeeping_kicking,goalkeeping_positioning,goalkeeping_reflexes,wage_eur


#### <b><i>Conclusión</i></b>
Según el dataset proporcionado y la manipulación de código llevada a cabo, se llega a la conclusión de que el dataset no posee ninguna fila duplicada en todoo su contenido.

#### <b><i>Decisión</i></b>
No se modificará el dataset.

##### <b>C. Valores incoherentes</b>

In [5]:
# C. Valores incoherentes
# C.1. Valores negativos
numeric_columns = [
    "age",
    "height_cm",
    "weight_kg",
    "overall",
    "potential",
    "attacking_crossing",
    "attacking_finishing",
    "attacking_heading_accuracy",
    "attacking_short_passing",
    "attacking_volleys",
    "skill_dribbling",
    "skill_curve",
    "skill_fk_accuracy",
    "skill_long_passing",
    "skill_ball_control",
    "movement_acceleration",
    "movement_sprint_speed",
    "movement_agility",
    "movement_reactions",
    "movement_balance",
    "defending_standing_tackle",
    "defending_sliding_tackle",
    "goalkeeping_diving",
    "goalkeeping_handling",
    "goalkeeping_kicking",
    "goalkeeping_positioning",
    "goalkeeping_reflexes",
    "wage_eur"
]
negatives = df[df[numeric_columns] < 0]

#Buscamos elementos NO nulos
negatives = pd.DataFrame({
    "Negativos": negatives.count()
})
negatives

,Negativos
age,0
height_cm,0
weight_kg,0
nationality_name,0
overall,0
potential,0
attacking_crossing,0
attacking_finishing,0
attacking_heading_accuracy,0
attacking_short_passing,0


#### <b><i>Conclusión</i></b>
Según el dataset proporcionado y la manipulación de código llevada a cabo, se llega a la conclusión de que el dataset no posee ningún elemento negativo que sea completamente incoherente con lo medido.

#### <b><i>Decisión</i></b>
No se modificará el dataset.

In [6]:
# C.2. Cardinalidad y unicidad de variable categórica
all_countries = df["nationality_name"].value_counts()
unique_countries = all_countries[all_countries == 1]
print("Nacionalidades que aparecen una vez: ", len(unique_countries))
unique_countries

Nacionalidades que aparecen una vez:  20


nationality_name
Tanzania            1
Fiji                1
Bermuda             1
Eritrea             1
Mauritius           1
Guam                1
Kyrgyzstan          1
Papua New Guinea    1
Ethiopia            1
Andorra             1
Korea DPR           1
Saint Lucia         1
Afghanistan         1
Vietnam             1
Belize              1
Bhutan              1
Barbados            1
Gibraltar           1
Estonia             1
Indonesia           1
Name: count, dtype: int64

In [7]:
not_unique_countries = all_countries[all_countries > 1]
print("Nacionalidades que aparecen más una vez: ", len(not_unique_countries))
not_unique_countries

Nacionalidades que aparecen más una vez:  143


nationality_name
England      1719
Germany      1214
Spain        1086
France        980
Argentina     960
             ... 
Lebanon         2
Guatemala       2
Palestine       2
Malaysia        2
Hong Kong       2
Name: count, Length: 143, dtype: int64

#### <b><i>Conclusión</i></b>
Según el dataset proporcionado y la manipulación de código llevada a cabo, se llega a la conclusión de que el dataset no posee ningún elemento incoherente con respecto a la variable categórica "nationality_name". Esto debido a que no se presentan países escritos con intercalo de letras mayúsculas y minúsculas. Es decir, no aparecen como datos únicos "Argentina" y "ARGENTINA", por ejemplo.
Asimismo, podemos observar la cardinalidad de todos los países (cuántas veces aparece cada nacionalidad). Esta cardinalidad nos permite decidir a qué países ponerle foco principalmente y hacer el estudio únicamente con ellos.

#### <b><i>Decisión</i></b>
Para facilitar nuestro estudios y reducir el número de la muestra por visibilidad gráfica, nos quedaremos con el top 20 países con más jugadores. La decisión viene respaldada del siguiente script, donde se encuentra como resultado que el 63.8% de todos los jugadores se encuentra en el top 20 de países con más jugadores.

In [8]:
# Obtenemos el top 20
top_twenty_countries = all_countries[:20]
percentage = (top_twenty_countries.sum()/all_countries.sum())*100
print(f"El porcentaje de jugadores que se encuentra en el top 20 es: {percentage}%")
top_twenty_countries

El porcentaje de jugadores que se encuentra en el top 20 es: 63.896130983418495%


nationality_name
England                1719
Germany                1214
Spain                  1086
France                  980
Argentina               960
Brazil                  897
Japan                   546
Netherlands             439
United States           413
Poland                  401
China PR                385
Sweden                  384
Norway                  379
Portugal                373
Republic of Ireland     373
Saudi Arabia            355
Mexico                  352
Italy                   338
Romania                 335
Korea Republic          325
Name: count, dtype: int64

In [9]:
# C.3. Variables outliers (solo para numéricas)
"""
    Acotación: utilizaremos dos arrays, uno que será de atributos de jugadores de campo;
    otro, para arqueros y otro de atributo generales. Esto debido a que podemos detectar 
    como outliers atributos de jugadores de campo para arqueros, y esto no nos daría un 
    análisis concreto y bien hecho.
"""
# C.3.1. Jugadores: Creamos la lista
numeric_columns_footballer = [
    "attacking_crossing",
    "attacking_finishing",
    "attacking_heading_accuracy",
    "attacking_short_passing",
    "attacking_volleys",
    "skill_dribbling",
    "skill_curve",
    "skill_fk_accuracy",
    "skill_long_passing",
    "skill_ball_control",
    "movement_acceleration",
    "movement_sprint_speed",
    "movement_agility",
    "movement_reactions",
    "movement_balance",
    "defending_standing_tackle",
    "defending_sliding_tackle",
]
# Utilizamos cuantiles 1 y 3
q1 = df[numeric_columns_footballer].quantile(0.25)
q3 = df[numeric_columns_footballer].quantile(0.75)

# Distancia Intercuartil
iqr = q3 - q1

# Límites
limite_inf = q1 - 1.5*iqr
limite_sup = q3 + 1.5*iqr

# Filtro
outliers_footballer = df[((df[numeric_columns_footballer] < limite_inf) | (df[numeric_columns_footballer] > limite_sup)).any(axis=1)]
print(outliers_footballer.shape)
outliers_footballer.head()

(2655, 29)


,age,height_cm,weight_kg,nationality_name,overall,potential,attacking_crossing,attacking_finishing,attacking_heading_accuracy,attacking_short_passing,...,movement_reactions,movement_balance,defending_standing_tackle,defending_sliding_tackle,goalkeeping_diving,goalkeeping_handling,goalkeeping_kicking,goalkeeping_positioning,goalkeeping_reflexes,wage_eur
0,34,170,72,Argentina,93,93,85,95,70,91,...,94,95,35,24,6,11,15,14,8,320000.0
1,32,185,81,Poland,92,92,71,95,90,85,...,93,82,42,19,15,6,12,8,10,270000.0
2,36,187,83,Portugal,91,91,87,95,90,80,...,94,74,32,24,7,11,15,14,11,270000.0
3,29,175,68,Brazil,91,91,85,83,63,86,...,89,84,32,29,9,9,15,15,11,270000.0
4,30,181,70,Belgium,91,91,94,82,55,94,...,91,78,65,53,15,13,5,10,13,350000.0


<b>Explicación</b>: Lo que hemos obtenido aquí es a <b>TODOS LOS JUGADORES QUE TENGAN POR LO MENOS UN ATRIBUTO OUTLIER SEGÚN LOS ATRIBUTOS QUE CORRESPONDAN A JUGADOR DE CAMPO (NO PORTERO)</b>.

Tengamos en cuenta nuestro contexto: Para poder hacer un buen análisis, necesitamos saber los valores outliers si o si. Sin embargo, debemos saber qué valores outliers tenemos también, es decir, su valor exacto. Ya sabemos, por el análisis anterior en C.1. que no tenemos negativos. Está bien, pero por ejemplo, podríamos tener a jugadores con edad 1, o podríamos tener a jugadores con edad 60. Técnicamente, no es imposible, pero claramente la probabilidad es ridículamente baja. Entonces, ese tipo de valores, son outliers sin sentido, que se deben remover. De igual forma, si se encontrara a un jugador con más de la mitad de atributos como outlier, se pensaría directamente que no es de este planeta, y sería muy raro, entonces, ese tipo de datos, también se borran. 

Sin embargo, existen excepcione, donde por ejemplo, un jugador podría tener 7, 8 outliers. Un ejemplo, Lionel Messi, con un `wage_eur` demasiado alto, entre otras habilidades. U otros jugadores como Kylian Mbappé, o Lamine Yamal. Tenemos que tener cuidado ya que son ESOS CASOS los que nos gustaría analizar y son los que deseamos predecir y entender QUÉ FACTORES realmente influyen en que esos jugadores tengan un salario tan alto.

In [10]:
#Ahora, identifiquemos la cantidad de outliers por cada jugador que tenga al menos 1 outlier
condicion_outliers = df[numeric_columns_footballer].lt(limite_inf, axis=1) | df[numeric_columns_footballer].gt(limite_sup, axis=1)
df["cantidad_outliers"] = condicion_outliers.sum(axis=1)
outliers_footballer = df[condicion_outliers.any(axis=1)]
print("La cantidad total de filas con al menos 1 outlier es:", len(outliers_footballer))
print("Porcentaje representativo: ", (len(outliers_footballer)/len(df))*100, "%")
outliers_footballer

La cantidad total de filas con al menos 1 outlier es: 2655
Porcentaje representativo:  13.843987902805296 %


,age,height_cm,weight_kg,nationality_name,overall,potential,attacking_crossing,attacking_finishing,attacking_heading_accuracy,attacking_short_passing,...,movement_balance,defending_standing_tackle,defending_sliding_tackle,goalkeeping_diving,goalkeeping_handling,goalkeeping_kicking,goalkeeping_positioning,goalkeeping_reflexes,wage_eur,cantidad_outliers
0,34,170,72,Argentina,93,93,85,95,70,91,...,95,35,24,6,11,15,14,8,320000.0,5
1,32,185,81,Poland,92,92,71,95,90,85,...,82,42,19,15,6,12,8,10,270000.0,1
2,36,187,83,Portugal,91,91,87,95,90,80,...,74,32,24,7,11,15,14,11,270000.0,1
3,29,175,68,Brazil,91,91,85,83,63,86,...,84,32,29,9,9,15,15,11,270000.0,2
4,30,181,70,Belgium,91,91,94,82,55,94,...,78,65,53,15,13,5,10,13,350000.0,3
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
19164,21,175,77,India,47,55,11,7,10,26,...,41,12,12,48,45,45,46,47,500.0,5
19166,18,186,74,India,47,67,13,8,10,22,...,44,14,13,49,41,39,45,49,500.0,7
19167,19,161,63,India,47,67,29,50,34,43,...,71,16,11,8,6,14,11,14,500.0,1
19170,22,181,74,India,47,55,13,8,13,20,...,43,13,13,49,46,43,47,52,500.0,7


Hemos encontrado que <b>2655 jugadores tienen al menos un outlier.</b> Esto representa un <b>13.84% de los 19178 jugadores incluidos en el dataset.</b> No es un procentaje alto, de hecho, representa una pequeña parte del total de los jugadores. Sin embargo, son los casos que más relevancia tienen en nuestro estudio. Este tipo de casos nos permitirán entender por qué Messi es el jugador mejor pagado del planeta, o por qué otros supercracks mundiales ganan tanto.

Entonces, no me interesará mucho saber la cantidad de jugadores que cuentan con estos valores outliers, sino cuál es el porcentaje de outliers en cada columna para poder considerar los atributos más importantes.

In [11]:
# Calculemos el porcentaje de outliers por columna
percentages = (condicion_outliers.sum()/len(df))*100
percentages

attacking_crossing             0.000000
attacking_finishing            0.000000
attacking_heading_accuracy     6.199812
attacking_short_passing        9.344040
attacking_volleys              0.000000
skill_dribbling               11.007404
skill_curve                    0.000000
skill_fk_accuracy              0.015643
skill_long_passing             0.297216
skill_ball_control            11.507978
movement_acceleration          3.373657
movement_sprint_speed          4.692877
movement_agility               1.334863
movement_reactions             1.861508
movement_balance               1.559078
defending_standing_tackle      0.000000
defending_sliding_tackle       0.000000
dtype: float64

In [12]:
# Obtenemos los 3 porcentajes más altos
ordered_percentages = percentages.sort_values() # sort_values ordena de menor a mayor
ordered_percentages[-3:]

attacking_short_passing     9.344040
skill_dribbling            11.007404
skill_ball_control         11.507978
dtype: float64

Hemos obtenido todos los porcentajes, y también hemos determinado cuáles son los 3 más altos. Es así, que podemos decir que los atributos `attacking_short_passing`, `skill_dribbling` y `skill_ball_control` son aquellos que tienen el mayor número de outliers. Por lo tanto, serán elementos diferenciadores en el análisis y la predicción del sueldos de los futbolistas.

Con respecto a la limpieza de datos, dado que en loss otros atributos no se han identificado porcentajes lo suficientemente significativos ni valores completamente extraño, solo un poco fuera del rango intercuartil, entonces, no se limpiará ninguna fila outlier.

#### <b><i>Decisión</i></b>
Dado el análisis realizado y la obtención de los valores outliers, no se determinaron valores lo suficientemente alejados de la realidad como para tomar la decisión de eliminar alguna fila. Por lo tanto, no se borrará dado este motivo. Además, la evidencia hallada resulta fundamental para la futura predicción de sueldos.

In [20]:
# C.3.2. Portero: Creamos la lista
numeric_columns_gk = [
    "goalkeeping_diving",
    "goalkeeping_handling",
    "goalkeeping_kicking",
    "goalkeeping_positioning",
    "goalkeeping_reflexes",
]

# Utilizamos cuantiles 1 y 3
q1 = df[numeric_columns_gk].quantile(0.25)
q3 = df[numeric_columns_gk].quantile(0.75)

# Distancia Intercuartil
iqr = q3 - q1

# Límites
limite_inf_gk = q1 - 1.5*iqr
limite_sup_gk = q3 + 1.5*iqr

# Filtro
outliers_gk = df[((df[numeric_columns_gk] < limite_inf_gk) | (df[numeric_columns_gk] > limite_sup_gk)).any(axis=1)]
print(outliers_gk.shape)
outliers_gk.head().drop(columns=["cantidad_outliers", "cantidad_outliers_common"])

(2132, 32)


,age,height_cm,weight_kg,nationality_name,overall,potential,attacking_crossing,attacking_finishing,attacking_heading_accuracy,attacking_short_passing,...,movement_balance,defending_standing_tackle,defending_sliding_tackle,goalkeeping_diving,goalkeeping_handling,goalkeeping_kicking,goalkeeping_positioning,goalkeeping_reflexes,wage_eur,cantidad_outliers_gk
5,28,188,87,Slovenia,91,93,13,11,15,43,...,49,12,18,87,92,78,90,90,130000.0,5
7,35,193,93,Germany,90,90,15,13,25,60,...,35,10,11,88,88,91,89,88,86000.0,5
8,29,187,85,Germany,90,92,18,14,11,61,...,43,13,10,88,85,88,88,90,250000.0,5
12,29,199,96,Belgium,89,91,14,14,13,33,...,45,18,16,84,89,74,86,88,250000.0,5
18,27,188,86,Brazil,89,91,20,14,14,61,...,48,15,8,87,82,93,88,88,200000.0,5


Al igual que con los jugadores, para los porteros se hace un análisis separado dado que las habilidades de estos difieren de las de los jugadores de campo. Es decir, las habilidades del portero están siendo aisladas para un estudio más compacto y preciso. Procederemos a hacer un análisis similar de igual forma.

In [21]:
#Ahora, identifiquemos la cantidad de outliers por cada jugador que tenga al menos 1 outlier de los atributos de portero
condicion_outliers_gk = df[numeric_columns_gk].lt(limite_inf_gk, axis=1) | df[numeric_columns_gk].gt(limite_sup_gk, axis=1)
df["cantidad_outliers_gk"] = condicion_outliers_gk.sum(axis=1)
outliers_gk = df[condicion_outliers_gk.any(axis=1)]
print("La cantidad total de filas con al menos 1 outlier es:", len(outliers_gk))
print("Porcentaje representativo: ", (len(outliers_gk)/len(df))*100, "%")
outliers_gk.drop(columns=["cantidad_outliers"])

La cantidad total de filas con al menos 1 outlier es: 2132
Porcentaje representativo:  11.1169047867348 %


,age,height_cm,weight_kg,nationality_name,overall,potential,attacking_crossing,attacking_finishing,attacking_heading_accuracy,attacking_short_passing,...,defending_standing_tackle,defending_sliding_tackle,goalkeeping_diving,goalkeeping_handling,goalkeeping_kicking,goalkeeping_positioning,goalkeeping_reflexes,wage_eur,cantidad_outliers_gk,cantidad_outliers_common
5,28,188,87,Slovenia,91,93,13,11,15,43,...,12,18,87,92,78,90,90,130000.0,5,3
7,35,193,93,Germany,90,90,15,13,25,60,...,10,11,88,88,91,89,88,86000.0,5,3
8,29,187,85,Germany,90,92,18,14,11,61,...,13,10,88,85,88,88,90,250000.0,5,3
12,29,199,96,Belgium,89,91,14,14,13,33,...,18,16,84,89,74,86,88,250000.0,5,4
18,27,188,86,Brazil,89,91,20,14,14,61,...,15,8,87,82,93,88,88,200000.0,5,3
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
19149,20,182,70,China PR,48,54,13,8,12,14,...,15,16,46,49,51,48,51,800.0,5,1
19155,19,180,72,England,48,65,12,5,12,29,...,12,13,52,49,49,43,48,1000.0,5,0
19164,21,175,77,India,47,55,11,7,10,26,...,12,12,48,45,45,46,47,500.0,5,1
19166,18,186,74,India,47,67,13,8,10,22,...,14,13,49,41,39,45,49,500.0,5,1


Al igual que en el ejercicio anterior, lo que nos interesa es conocer el porcentaje de outliers por cada atributo correspondiente a los arqueros, por lo tanto, en el siguiente script, calcularemos dichos porcentajes.

In [22]:
percentages_gk = (condicion_outliers_gk.sum()/len(df))*100
percentages_gk

goalkeeping_diving         11.090833
goalkeeping_handling       11.090833
goalkeeping_kicking        11.111690
goalkeeping_positioning    11.090833
goalkeeping_reflexes       11.096048
dtype: float64

Podemos observar que practicamente todos los atributos correspondientes a los arqueros tienen un porcentaje de outliers similar. Entonces, resulta relevante analizar cada variable correspondiente a los arqueros en nuestro estudio.

#### <b><i>Decisión</i></b>
Dado el análisis realizado y la obtención de los valores outliers, no se determinaron valores lo suficientemente alejados de la realidad como para tomar la decisión de eliminar alguna fila. Por lo tanto, no se borrará dado este motivo. Además, la evidencia hallada resulta fundamental para la futura predicción de sueldos.

In [23]:
# C.3.3. Valores comunes (age, height_cm, weight_kg, overall, potential y wage_eur)
common_values = [
    "age",
    "height_cm",
    "weight_kg",
    "overall",
    "potential",
    "wage_eur"
]

# Utilizamos los cuantiles q1 y q3
q1 = df[common_values].quantile(0.25)
q3 = df[common_values].quantile(0.75)

# Distancia Intercuartil
iqr = q3 - q1

# Límites
limite_inf_common = q1 - 1.5*iqr
limite_sup_common = q3 + 1.5*iqr

# Outliers
outlier_common = df[((df[common_values] < limite_inf_common) | (df[common_values] > limite_sup_common)).any(axis=1)]
print(outlier_common.shape)
outlier_common.drop(columns = ["cantidad_outliers", "cantidad_outliers_gk"])

(2570, 32)


,age,height_cm,weight_kg,nationality_name,overall,potential,attacking_crossing,attacking_finishing,attacking_heading_accuracy,attacking_short_passing,...,movement_balance,defending_standing_tackle,defending_sliding_tackle,goalkeeping_diving,goalkeeping_handling,goalkeeping_kicking,goalkeeping_positioning,goalkeeping_reflexes,wage_eur,cantidad_outliers_common
0,34,170,72,Argentina,93,93,85,95,70,91,...,95,35,24,6,11,15,14,8,320000.0,3
1,32,185,81,Poland,92,92,71,95,90,85,...,82,42,19,15,6,12,8,10,270000.0,3
2,36,187,83,Portugal,91,91,87,95,90,80,...,74,32,24,7,11,15,14,11,270000.0,3
3,29,175,68,Brazil,91,91,85,83,63,86,...,84,32,29,9,9,15,15,11,270000.0,3
4,30,181,70,Belgium,91,91,94,82,55,94,...,78,65,53,15,13,5,10,13,350000.0,3
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
19173,22,180,64,China PR,47,52,46,32,48,50,...,70,43,48,6,10,5,15,13,1000.0,2
19174,19,175,70,Republic of Ireland,47,59,54,33,46,51,...,69,44,47,11,12,6,8,10,500.0,1
19175,21,178,72,Republic of Ireland,47,55,39,32,43,49,...,59,44,48,8,6,7,10,6,500.0,1
19176,19,173,66,Republic of Ireland,47,60,29,49,40,38,...,73,14,11,7,10,7,14,15,500.0,1


En este último estudio estamos tomando los atributos comunes de todos (sea portero, sea jugador). El análisis, entonces, resulta ser de 3 bloques: atributos comunes, atributos de portero y atributos de jugador de campo. Procedemos de la misma manera que los anteriores casos.

In [24]:
#Ahora, identifiquemos la cantidad de outliers por cada jugador que tenga al menos 1 outlier de los atributos comunes
condicion_outliers_common = df[common_values].lt(limite_inf_common, axis=1) | df[common_values].gt(limite_sup_common, axis=1)
df["cantidad_outliers_common"] = condicion_outliers_common.sum(axis=1)
outliers_common = df[condicion_outliers_common.any(axis=1)]
print("La cantidad total de filas con al menos 1 outlier es:", len(outliers_common))
print("Porcentaje representativo: ", (len(outliers_common)/len(df))*100, "%")
outliers_common.drop(columns=["cantidad_outliers", "cantidad_outliers_gk"])

La cantidad total de filas con al menos 1 outlier es: 2570
Porcentaje representativo:  13.400771717593074 %


,age,height_cm,weight_kg,nationality_name,overall,potential,attacking_crossing,attacking_finishing,attacking_heading_accuracy,attacking_short_passing,...,movement_balance,defending_standing_tackle,defending_sliding_tackle,goalkeeping_diving,goalkeeping_handling,goalkeeping_kicking,goalkeeping_positioning,goalkeeping_reflexes,wage_eur,cantidad_outliers_common
0,34,170,72,Argentina,93,93,85,95,70,91,...,95,35,24,6,11,15,14,8,320000.0,3
1,32,185,81,Poland,92,92,71,95,90,85,...,82,42,19,15,6,12,8,10,270000.0,3
2,36,187,83,Portugal,91,91,87,95,90,80,...,74,32,24,7,11,15,14,11,270000.0,3
3,29,175,68,Brazil,91,91,85,83,63,86,...,84,32,29,9,9,15,15,11,270000.0,3
4,30,181,70,Belgium,91,91,94,82,55,94,...,78,65,53,15,13,5,10,13,350000.0,3
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
19173,22,180,64,China PR,47,52,46,32,48,50,...,70,43,48,6,10,5,15,13,1000.0,2
19174,19,175,70,Republic of Ireland,47,59,54,33,46,51,...,69,44,47,11,12,6,8,10,500.0,1
19175,21,178,72,Republic of Ireland,47,55,39,32,43,49,...,59,44,48,8,6,7,10,6,500.0,1
19176,19,173,66,Republic of Ireland,47,60,29,49,40,38,...,73,14,11,7,10,7,14,15,500.0,1


In [25]:
percentages_common = (condicion_outliers_common.sum()/len(df))*100
percentages_common

age           0.031286
height_cm     0.140786
weight_kg     0.344144
overall       0.829075
potential     0.834289
wage_eur     12.488268
dtype: float64

Podemos observar, que en su mayoría, las variables comunes a todos están normales, no es nada del otro mundo. Lo que nos dice que este dataset lleno de jugadores, en promedio, se distribuyen todos de la misma manera. Sin embargo, la mayor diferencia la encontramos en la variable target `wage_eur`, con un 12.5% de outliers con respecto al total. Eso quiere decir que nuestro objetivo será explicar por qué existe ese porcentaje de outliers en nuestra variable target.

### <b> CONCLUSIÓN FINAL DEL BLOQUE </b>
Según el estudio realizado al dataset, los hallazgos nos dicen la mayoría de datos <b>son relevantes para nuestro estudio</b> debido a que no existen valores duplicados, missing ni outliers sin sentido. Es más, los outliers son la parte más interesante del estudio debido a que la mayor cantidad de estos podemos encontrarla en la variable `wage_eur` que es nuestra variable target y la que intentaremos predecir mediante regresión. Es así que se tomará como decisión <b>no eliminar ninguna fila del dataset y conservarlo tal cual fue entregado</b>.

En la siguiente parte de nuestro estudio exploraremos todos los datos que hemos tomado, analizaremos gráficos y tomaremos decisiones para poder seguir mejorando la calidad de nuestro estudio.